nbqa/replace_source.py

```diff
@@ -5,20 +5,50 @@
 """
 
 import json
-from typing import TYPE_CHECKING, Dict, List
+from typing import TYPE_CHECKING, Iterator, List, Mapping, Set
 
+from nbqa.notebook_info import NotebookInfo
 from nbqa.save_source import CODE_SEPARATOR
 
 if TYPE_CHECKING:
     from pathlib import Path
 
 
-def main(
-    python_file: "Path",
-    notebook: "Path",
-    trailing_semicolons: List[int],
-    temporary_lines: Dict[str, str],
-) -> None:
+def _reinstate_magics(
+    source: str,
+    cell_number: int,
+    trailing_semicolons: Set[int],
+    temporary_lines: Mapping[str, str],
+) -> List[str]:
+    """
+    Put (preprocessed) line magics back in.
+
+    Parameters
+    ----------
+    source
+        Portion of Python file between cell separators.
+    cell_number
+        Number of current cell.
+    trailing_semicolons
+        List of cells which originally had trailing semicolons.
+    temporary_lines
+        Mapping from temporary lines to original lines.
+
+    Returns
+    -------
+    List[str]
+        New source that can be saved into Jupyter Notebook.
+    """
+    rstripped_source = source.rstrip()
+    if cell_number in trailing_semicolons and not rstripped_source.endswith(";"):
+        source = rstripped_source + ";"
+    # we take [1:] because the first cell is just '\n'
+    for key, val in temporary_lines.items():
+        source = source.replace(key, val)
```

Comment:
Oops, looks like I put the comment  on the wrong line in a previous PR

```suggestion
    for key, val in temporary_lines.items():
        source = source.replace(key, val)
    # we take [1:] because the first cell is just '\n'
```